sched/signal/sig_dispatch.c

```diff
@@ -373,11 +373,11 @@ int nxsig_tcbdispatch(FAR struct tcb_s *stcb, siginfo_t *info)
 #endif
     {
 #ifdef CONFIG_LIB_SYSCALL
-      /* If the thread is in syscall, schedule the sigaction here */
+      /* If the thread is in syscall, queue the sigaction here */
 
-      if ((stcb->flags & TCB_FLAG_SYSCALL) != 0)
+      if (masked == 0)
```

Comment:
why not merge with line 425-430?

Reply:
I don't think that would be logically equivalent.

The first if has really only one significant test when a signal is received during as system call:

```
if (stcb->task_state == TSTATE_WAIT_SIG &&
          (masked == 0 || nxsig_ismember(&stcb->sigwaitmask, info->si_signo)))
```

If masked == 0 and `stcb->task_state == TSTATE_WAIT_SIG` is true, the second branch will not be executed. So this has to be done either here, or moved inside the first if. I think this is better as we don't do the work of `nxsig_queue_action` inside a critical section.

Also, the else-branch must test for masked==1, otherwise the signal is delivered twice (again). I agree this is a bit messy but I cannot see a better way to do this.

Reply:
I still can't understand why can't move after sigwait process.

Reply:
Very good question. I frankly do not have a 100% clear picture how this is supposed to work. So you suggest that:
- The waiting task is woken up, possibly causing a context switch

When the signaling task is scheduled back:
- The signal action is queued after waking up the waiting task, if the signal is not masked
- The signal is pended, if the signal is masked
?

Btw; the original logic was modified here: https://github.com/apache/nuttx/pull/2851

As a regression, kill() / SIGKILL / SIGTERM no longer works if the task/process is inside a system call AND waiting for a signal (sigwait, sleep, etc). As the task is not waiting for SIGKILL/SIGTERM, it was put into the pending queue (else-branch) and the task was correctly terminated. This is what I noticed and am trying to fix.

Reply:
Here is my thought:

1. If mask == 1 and task is waiting for signal, wake up the task
2. Otherwise, call nxsig_add_pendingsignal regardless whether the task is/isn't in syscall.

Reply:
I think the original implementation worked like this, so this would mean to revert #2851 ?

What about the signal action in case 1. ? Even if a signal is masked, the task might have injected an action / signal handler for the signal. When the procsigmask changes I think the signal handler should be called, i.e. the signal action should be queued even in case 1. Or am I missing something ?

Reply:
Just to clarify a bit more, if you take a look into the unmasked signal handling, the first thing that is done is a signal *action* is queued. Then if the task is waiting for a signal it is also unblocked / woken up. This way the sigunbmask is received by sigwait() and if a signal action handler is installed by the user, it also gets called. This is logical to me.

https://github.com/apache/nuttx/blob/fce927123ed243426814f38b6e89b82ac859a7bc/sched/signal/sig_dispatch.c#L434-L475

The masked signal handling is not clear to me. The task is woken up by this, but no signal action is added anywhere. So the user handler is never called, even if the procmask changes.

https://github.com/apache/nuttx/blob/0e4466682888cf32ca661aa04c649a836860a22f/sched/signal/sig_dispatch.c#L389-L426

This is what I don't understand and I think there is an error here, i.e. depending on whether this is true / false, the signal action is properly queued or not, and the user handler is either called (when sigprocmask changes) or not.

```
      if (stcb->task_state == TSTATE_WAIT_SIG &&
          (masked == 0 ||
           nxsig_ismember(&stcb->sigwaitmask, info->si_signo)))
```

Please note that `nxsig_add_pendingsignal` both wakes up the task when sigprocmask changes _AND_ queues the signal action, meaning the signal handler is called also.

Reply:
> I think the original implementation worked like this, so this would mean to revert #2851 ?
> 

It have been done.

> What about the signal action in case 1. ? Even if a signal is masked, the task might have injected an action / signal handler for the signal or there might be a default action. When the procsigmask changes I think the signal handler should be called, i.e. the signal action should be queued even in case 1. Or am I missing something ?

Yes, the signal should be queue regardless the signal mask setting.

> Just to clarify a bit more, if you take a look into the unmasked signal handling, the first thing that is done is a signal _action_ is queued. Then if the task is waiting for a signal it is also unblocked / woken up. This way the sigunbmask is received by sigwait() and if a signal action handler is installed by the user, it also gets called. This is logical to me.
> 

Normally, the user either install the signal action and unmask the signal, or mask the signal but call sigwait, it's very rare to use both at the same time. Another method to receive the signal is signalfd which is added recently: https://github.com/apache/nuttx/pull/7991.

> https://github.com/apache/nuttx/blob/fce927123ed243426814f38b6e89b82ac859a7bc/sched/signal/sig_dispatch.c#L434-L475
> 
> The masked signal handling is not clear to me. The task is woken up by this, but no signal action is added anywhere. So the user handler is never called, even if the sigprocmask changes.
> 
> https://github.com/apache/nuttx/blob/0e4466682888cf32ca661aa04c649a836860a22f/sched/signal/sig_dispatch.c#L389-L426
> 
> This is what I don't understand and I think there is an error here, i.e. depending on whether this is true / false, the signal action is properly queued or not, and the user handler is either called (when sigprocmask changes) or not.
> 

No, it's right behaviour. We don't need do any additional thing except return directly since it's the expected by the caller of sigwait: https://man7.org/linux/man-pages/man3/sigwait.3.html

As, I mention before, it's seldom that user call sigwait/signal action/signalfd at the same time. 

> ```
>       if (stcb->task_state == TSTATE_WAIT_SIG &&
>           (masked == 0 ||
>            nxsig_ismember(&stcb->sigwaitmask, info->si_signo)))
> ```
> 
> Please note that `nxsig_add_pendingsignal` is needed in this case to deliver the signal action if/when the sigprocmask changes.
> 
> This is the odd part to me, this branch just wakes the task up but the signal handler is never called, regardless of any system call involvement.

Reply:
>As, I mention before, it's seldom that user call sigwait/signal action/signalfd at the same time.

I agree but what if there is a default sigaction, like SIGTERM. If the user app for example is written like this:
```

int main(void)
{
  while(true)
    {
      sleep(5);
    }
  return 0;
}
```

It will be in sleep() which means it is waiting for a signal. Then sending `kill *pid*` will simply wake the task but not kill it. This works in flat mode / when system calls are not enabled as SIGTERM can never be masked, so the masked==0 branch is executed. But when system calls are enabled the SIGTERM action is never executed :(

This is the issue I face, but I will fix it next week by making a new queue for pended signal actions when the task is in system call. Using nxsig_add_pendingsignal  is not correct for this as it also adds a pending signal which will abort any subsequent call to sigwait() and its friends.

Reply:
Btw thank you for the comment, and thank you for your hard work

Reply:
Now I finally understand, I did not read documentation of sigwait() carecully enough. 

> The function accepts the signal (removes it from the pending list of signals), and returns the signal number in sig.

So the only thing that needs to be done is to fix the system call behavior, when masked == 0. The signal action is not currently performed. I'll provide a patch in a moment.